prog/hints.go

```diff
@@ -0,0 +1,343 @@
+package prog
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/google/syzkaller/sys"
+)
+
+// A set of calls for which hints will not be generated
+var hintNamesBlackList = map[string]bool{
+	"mmap":  true,
+	"open":  true,
+	"close": true,
+}
+
+// Size in bytes for each type of comparison exported from KCOV
+var sizeMap = map[KcovComparisonType]uint{
+	KCOV_TYPE_CMP1:       1,
+	KCOV_TYPE_CMP2:       2,
+	KCOV_TYPE_CMP4:       4,
+	KCOV_TYPE_CMP8:       8,
+	KCOV_TYPE_SWITCH1:    1,
+	KCOV_TYPE_SWITCH2:    2,
+	KCOV_TYPE_SWITCH4:    4,
+	KCOV_TYPE_SWITCH8:    8,
+	KCOV_TYPE_CONST_CMP1: 1,
+	KCOV_TYPE_CONST_CMP2: 2,
+	KCOV_TYPE_CONST_CMP4: 4,
+	KCOV_TYPE_CONST_CMP8: 8,
+}
+
+// A set of special values for which the hints should not be created
+// Taken from prog/rand.go specialInts array
+// Created as a global constant map for performance reasons
+var specialIntsMap = map[uintptr]bool{
+	0:             true,
+	1:             true,
+	31:            true,
+	32:            true,
+	63:            true,
+	64:            true,
+	127:           true,
+	128:           true,
+	129:           true,
+	255:           true,
+	256:           true,
+	257:           true,
+	511:           true,
+	512:           true,
+	1023:          true,
+	1024:          true,
+	1025:          true,
+	2047:          true,
+	2048:          true,
+	4095:          true,
+	4096:          true,
+	(1 << 15) - 1: true,
+	(1 << 15):     true,
+	(1 << 15) + 1: true,
+	(1 << 16) - 1: true,
+	(1 << 16):     true,
+	(1 << 16) + 1: true,
+	(1 << 31) - 1: true,
+	(1 << 31):     true,
+	(1 << 31) + 1: true,
+	(1 << 32) - 1: true,
+	(1 << 32):     true,
+	(1 << 32) + 1: true,
+}
+
+type ArgKey struct {
+	value uintptr // all the values are promoted to uintptr
+	size  uint
+}
+type ArgVal struct {
+	// array of indices to locate the value (array needed to handle nested data)
+	// e.g. if the value is stored in:
+	// the third field of a struct, which is
+	// the fifth element of an array, which is
+	// the second argument of a syscall
+	// then the array will look like: {1, 5, 2}
+	indices []uint
+}
+
+// Maps of this type are used to store the arguments' values for each
+// executed syscall
+type ArgMap map[ArgKey][]*ArgVal
+
+// Casts an uintptr value to an integer of given size
+// size is given in bytes
+func castToSize(val uintptr, size uint) uintptr {
```

Comment:
Generally in Go prefer to place the most important functions at the top of the file and the less important functions closer to the bottom. The most important here are: MutateWithHint and GenerateHints, because they are part of public interface, so if a reader opens this file he comes for these functions.
In C/C++ it's common to dump tons of uninteresting stuff onto a reader first, but that's only because the language requires forward declarations to save one compiler pass.